bots/lending-liquidator/src/liquidate.ts

```diff
@@ -0,0 +1,163 @@
+        let result: [MonetaryAmount<CurrencyExt>, CurrencyExt, AccountId] | undefined;
+        undercollateralizedBorrowers.forEach((position) => {
+            // Among the collateral currencies locked by this borrower, which one is worth the most?
+            const highestValueCollateral = findHighestValueCollateral(position.collateralPositions, oracleRates); 
+            if (!highestValueCollateral) {
+                return;
+            }
+            // Among the borrowed currencies of this borrower, which one accepts the biggest repayment?
+            // (i.e. is worth the most after considering the close factor)
+            position.borrowPositions.forEach((loan) => {
+                const totalDebt = loan.amount.add(loan.accumulatedDebt);
+                // If this loan can be repaid using the lending-liquidator's balance
+                if (liquidatorBalance.has(totalDebt.currency)) {
+                    const loansMarket = markets.get(totalDebt.currency) as LoansMarket;
+                    const closeFactor = decodePermill(loansMarket.closeFactor);
+                    const balance = liquidatorBalance.get(totalDebt.currency) as ChainBalance;
+                    const rate = oracleRates.get(totalDebt.currency) as ExchangeRate<Currency, CurrencyExt>;
+                    // Can only repay a fraction of the total debt, defined by the `closeFactor`
+                    const repayableAmount = totalDebt.mul(closeFactor).min(balance.free);
+                    const referenceRepayable = referenceValue(repayableAmount, rate).min(highestValueCollateral.referenceValue);
+                    if (referenceRepayable.gt(maxRepayableLoan)) {
+                        maxRepayableLoan = referenceRepayable;
+                        result = [repayableAmount, highestValueCollateral.collateral.currency, position.accountId];
+                    }
+                }
+            })
+        });
+        return result;
+    }
+    
+    async function checkForLiquidations(interBtcApi: InterBtcApi, chainAssets: Set<Currency>): Promise<void> {
+        const accountId = addressOrPairAsAccountId(interBtcApi.api, interBtcApi.account as AddressOrPair);
+        const liquidatorBalance: Map<Currency, ChainBalance> = new Map();
+        const oracleRates: Map<Currency, ExchangeRate<Currency, CurrencyExt>> = new Map();
+
+        const [_balancesPromise, _oraclePromise, undercollateralizedBorrowers, marketsArray, foreignAssets] = await Promise.all([
+            Promise.all([...chainAssets].map((asset) => interBtcApi.tokens.balance(asset, accountId).then((balance) => liquidatorBalance.set(asset, balance)))),
+            Promise.all([...chainAssets].map((asset) => interBtcApi.oracle.getExchangeRate(asset).then((rate) => oracleRates.set(asset, rate)))),
+            interBtcApi.loans.getUndercollateralizedBorrowers(),
+            interBtcApi.loans.getLoansMarkets(),
+            interBtcApi.assetRegistry.getForeignAssets()
```

Comment:
Here, I think foreign assets could be passed as parameter to `checkForLiquidations` so they are not fetched with each block as they don't change very frequently.

Reply:
They don't change very frequently but refreshing the assets would require restarting the bot, and the goal is to have this as a long-running service. The vault client uses a similar approach.